### ISSUE TITLE
Invalid Metal conversion for array of textures

### ISSUE THREAD

**Issue by gemorin** [KhronosGroup/SPIRV-Cross#507]
I have a glsl fragment shader that gets a texture2D array uniform: i.e ```layout(binding = 2) uniform texture2D shipTextures[2]``` that I have to compiled into spirv with glslc

The spirv-cross Metal conversion generates incorrect input args for the Metal shader: it basically ignores the fact it's an array array and generates a normal texture2d variable
```
fragment main0_out main0(main0_in in [[stage_in]], constant PushConsts& pushConsts [[buffer(0)]], texture2d<float> shipTextures [[texture(0)]], sampler s [[sampler(0)]])
```

However Metal supports array of textures (and samplers).

This patch fixes the problem for me:
```
--- a/spirv_msl.cpp
+++ b/spirv_msl.cpp
@@ -3091,7 +3091,15 @@ string CompilerMSL::entry_point_args(bool append_comma)
                case SPIRType::Image:
                        if (!ep_args.empty())
                                ep_args += ", ";
-                       ep_args += type_to_glsl(type, var_id) + " " + r.name;
+                       if (is_array(type)) {
+                               ep_args += " array<";
+                               ep_args += type_to_glsl(type, var_id) + ", " + convert_to_string(type.array[0]);
+                               ep_args += "> " + r.name;
+                       }
+                       else
+                       {
+                               ep_args += type_to_glsl(type, var_id) + " " + r.name;
+                       }
                        ep_args += " [[texture(" + convert_to_string(r.index) + ")]]";
                        break;
                default:
```
I am very new to this codebase (and Metal/Vulkan for that matter) so my apologies in advance if I missed something.  Note that this patch also does not handle array of samplers but a similar change can be applied in the case SPIRType::Sampler case.

With this patch, the Metal code is as expected:
```
fragment main0_out main0(main0_in in [[stage_in]], constant PushConsts& pushConsts [[buffer(0)]],  array<texture2d<float>, 2> shipTextures [[texture(0)]], sampler s [[sampler(0)]])
```

I am attaching the spirv code just in case 
[fragment.spv.gz](https://github.com/KhronosGroup/SPIRV-Cross/files/1817739/fragment.spv.gz)

HTH

**Comment by HansKristian-ARM** [comment]
I'll have a look after GDC.

**Comment by VinDuv** [comment]
I have encountered this problem with shaders generated by the Dolphin emulator, that use an array of `sampler2DArray`. This reduced test case:
```
#version 450 core
layout(set = 1, binding = 0) uniform sampler2DArray samp[8];
void main() {
	vec4 a = texture(samp[0], gl_FragCoord.xyz);
}
```
gets translated to
```
fragment void main0(texture2d_array<float> samp [[texture(0)]], sampler sampSmplr [[sampler(0)]], float4 gl_FragCoord [[position]]) {
    float4 a = samp[0].sample(samp[0]Smplr, gl_FragCoord.xyz.xy, uint(round(gl_FragCoord.xyz.z)));
}
```

Applying your patch fixes the `texture2d_array` part. I’ve attempted to fix the `sampler` part with the same logic (I’m also quite unfamiliar with Metal and Vulkan):
```diff
--- a/spirv_msl.cpp
+++ b/spirv_msl.cpp
@@ -3085,13 +3085,29 @@ string CompilerMSL::entry_point_args(bool append_comma)
                case SPIRType::Sampler:
                        if (!ep_args.empty())
                                ep_args += ", ";
-                       ep_args += "sampler " + r.name;
+                       if (is_array(type)) {
+                               ep_args += " array<";
+                               ep_args += "sampler, " + convert_to_string(type.array[0]);
+                               ep_args += "> " + r.name;
+                       }
+                       else
+                       {
+                               ep_args += "sampler " + r.name;
+                       }
                        ep_args += " [[sampler(" + convert_to_string(r.index) + ")]]";
```

That results in what I think is the correct declaration:
```
array<texture2d_array<float>, 8> samp [[texture(0)]],  array<sampler, 8> sampSmplr [[sampler(0)]]
```

However, at the point of use, the incorrect code `samp[0]Smplr` is still generated. The converter adds `Smplr` as a suffix, which doesn’t play well with the array notation.

As a workaround, I tried turning the suffix into a prefix:

```diff
--- a/spirv_msl.hpp
+++ b/spirv_msl.hpp
@@ -302,7 +302,7 @@ protected:
        std::string stage_in_var_name = "in";
        std::string stage_out_var_name = "out";
        std::string stage_uniform_var_name = "uniforms";
-       std::string sampler_name_suffix = "Smplr";
+       std::string sampler_name_prefix = "smplr_for_";
        spv::Op previous_instruction_opcode = spv::OpNop;
 
        // OpcodeHandler that handles several MSL preprocessing operations.
```
That seems to fix the shader compilation. I don’t know it it’s completely correct, however.

**Comment by HansKristian-ARM** [comment]
So, I've got a patch which should work here, but I'm running into compilation issues. Apparently `array<T, N>` does not exist on MSL even though the specification says it should. Can you provide a sample MSL shader which uses arrays of textures for reference?

**Comment by gemorin** [comment]
@HansKristian-ARM 
This is the full MSL shader produced by my patched SPIRV-Cross:
```
#include <metal_stdlib>
#include <simd/simd.h>

using namespace metal;

struct PushConsts
{
    char pad0[640];
    int idx[10];
    float easingFactor[10];
    float colorFix;
};

struct main0_in
{
    int instanceIdx [[user(locn2)]];
    float2 uv [[user(locn1)]];
};

struct main0_out
{
    float4 outColor [[color(0)]];
};

fragment main0_out main0(main0_in in [[stage_in]], constant PushConsts& pushConsts [[buffer(0)]],  array<texture2d<float>, 7> shipTextures [[texture(0)]], sampler s [[sampler(0)]])
{
    main0_out out = {};
    out.outColor = shipTextures[pushConsts.idx[in.instanceIdx]].sample(s, in.uv);
    out.outColor.x += ((pushConsts.colorFix * 0.3) * length(out.outColor.xyz));
    out.outColor.y -= ((pushConsts.colorFix * 0.4) * length(out.outColor.xyz));
    out.outColor.z -= ((pushConsts.colorFix * 0.4) * length(out.outColor.xyz));
    float easing = pushConsts.easingFactor[in.instanceIdx];
    out.outColor.w *= mix(0.0, 1.0, easing);
    return out;
}
```

I have just took a look at https://developer.apple.com/metal/Metal-Feature-Set-Tables.pdf and both array of textures and samplers are only available starting with iOS 10 and macOS 10.13

**Comment by billhollings** [comment]
@HansKristian-ARM  Thanks so much for tackling this.

>I have just took a look at https://developer.apple.com/metal/Metal-Feature-Set-Tables.pdf and both array of textures and samplers are only available starting with iOS 10 and macOS 10.13.

Yes...this corresponds to MSL v2.0 (it's actually iOS 11).

`CompilerMSL::msl_options::msl_version` is available to interpret the MSL version within SPIRV-Cross. So...we should really be testing for this in code that depends on features only available in newer MSL versions.
